File: survey/domain/src/main/java/me/nalab/survey/domain/survey/FormQuestionable.java
package me.nalab.survey.domain.survey;

import java.time.Instant;
import java.util.Comparator;
import java.util.function.LongSupplier;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import lombok.experimental.SuperBuilder;
import me.nalab.survey.domain.exception.IdAlreadyGeneratedException;
import me.nalab.survey.domain.support.IdGeneratable;
import me.nalab.survey.domain.survey.spi.QuestionFeedbackValidable;

@SuperBuilder
@Getter
@ToString
@EqualsAndHashCode
public abstract class FormQuestionable implements IdGeneratable, Comparable<FormQuestionable> {

	protected Long id;
	protected final String title;
	protected final Instant createdAt;
	protected final Instant updatedAt;
	protected final Integer order;
	protected final QuestionType questionType;

	@Override
	public void withId(LongSupplier idSupplier) {
		if(id != null) {
			throw new IdAlreadyGeneratedException(this);
		}
		this.id = idSupplier.getAsLong();
		cascadeId(idSupplier);
	}

	abstract void throwIfIsNotValidQuestionFeedback(QuestionFeedbackValidable questionFeedbackable);

	protected void cascadeId(LongSupplier idSupplier) {
	}

	@Override
	public int compareTo(FormQuestionable other) {
		return Comparator.comparingInt(FormQuestionable::getOrder)
			.compare(this, other);
	}
}

File: survey/src/main/kotlin/me/nalab/survey/domain/survey/FormQuestionable.kt
package me.nalab.survey.domain.survey

import me.nalab.core.data.common.TimeBaseEntity
import java.util.function.LongSupplier
import javax.persistence.*

@Entity
@Table(name = "form_question")
abstract class FormQuestionable(
    @Id
    @Column(name = "form_question_id")
    open var id: Long? = null,

    @Column(name = "title", nullable = false, length = 45)
    protected open val title: String,

    @Column(name = "orders", nullable = false)
    open val order: Int,

    @Enumerated(EnumType.STRING)
    @Column(name = "question_type")
    protected open val questionType: QuestionType,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "survey_id", nullable = false)
    open val survey: Survey,
) : Comparable<FormQuestionable>, TimeBaseEntity() {

    protected open fun cascadeId(idSupplier: LongSupplier) {
        return
    }

    override fun compareTo(other: FormQuestionable): Int {
        return Comparator.comparingInt { obj: FormQuestionable -> obj.order }
            .compare(this, other)
    }
}
